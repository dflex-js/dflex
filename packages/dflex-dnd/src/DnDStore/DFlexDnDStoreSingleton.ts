import { canUseDOM } from "@dflex/utils";
import DnDStoreImp from "./DFlexDnDStore";

declare global {
  // eslint-disable-next-line
  var $DFlex: DnDStoreImp;
}

export default (function createStoreInstance() {
  const store = new DnDStoreImp();

  if (__DEV__) {
    if (canUseDOM()) {
      if (!globalThis.$DFlex) {
        globalThis.$DFlex = store;
      } else {
        throw new Error("DFlex store instances is already defined.");
      }
    }
  }

  return store;
})();
