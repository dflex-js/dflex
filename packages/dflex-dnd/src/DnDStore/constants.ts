const ON_OUT_CONTAINER = "$onDragOutContainer";
const ON_OUT_THRESHOLD = "$onDragOutThreshold";
const ON_DRAG_OVER = "$onDragOver";
const ON_DRAG_LEAVE = "$onDragLeave";
const ON_LIFT_UP = "$onLiftUpSiblings";
const ON_MOVE_DOWN = "$onMoveDownSiblings";

export const DRAG_EVT = Object.freeze({
  ON_OUT_CONTAINER,
  ON_OUT_THRESHOLD,
});

export const INTERACTIVITY_EVT = Object.freeze({
  ON_DRAG_OVER,
  ON_DRAG_LEAVE,
});

export const SIBLINGS_EVT = Object.freeze({
  ON_LIFT_UP,
  ON_MOVE_DOWN,
});

export const DFLEX_EVENTS = Object.freeze({
  ...DRAG_EVT,
  ...INTERACTIVITY_EVT,
  ...SIBLINGS_EVT,
});
