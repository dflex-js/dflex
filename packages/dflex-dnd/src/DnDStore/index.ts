export { default as store } from "./DFlexDnDStore";
export { default as initDFlexEvent } from "./DFlexEvents";
export { default as initDFlexListeners } from "./DFlexListeners";

export type { ElmTree } from "./types";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexEventsMap,
  DFlexEventPlugin as DFlexEventInitializer,
} from "./DFlexEvents";

export type {
  LayoutState as DFlexLayoutState,
  DFlexLayoutStateEvent,
} from "./DFlexListeners";
