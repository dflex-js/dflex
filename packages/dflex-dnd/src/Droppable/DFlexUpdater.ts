/* eslint-disable no-param-reassign */
import { Node } from "@dflex/core-instance";
import type { INode } from "@dflex/core-instance";

import { Direction, IPointAxes, PointNum } from "@dflex/utils";
import type { IPointNum, Axis } from "@dflex/utils";

import type { InteractivityEvent } from "../types";
import type { IDraggableInteractive } from "../Draggable";

import store from "../DnDStore";

function emitInteractiveEvent(
  type: InteractivityEvent["type"],
  element: INode
) {
  const evt: InteractivityEvent = {
    id: element.id,
    index: element.order.self,
    target: element.ref!,
    timeStamp: Date.now(),
    type,
  };

  store.emitEvent(evt);
}

/**
 * Calculates the distance between two elements and update the targeted element
 * accordingly.
 */
class DFlexUpdater {
  protected draggable: IDraggableInteractive;

  private elmTransition: IPointNum;

  private draggedPositionOffset: IPointNum;

  private draggedTransition: IPointNum;

  private infiniteTransformCount: number;

  /** Isolated form the threshold and predict is-out based on the controllers */
  protected isParentLocked: boolean;

  constructor(draggable: IDraggableInteractive) {
    this.draggable = draggable;

    /**
     * Next element calculated transition space.
     */
    this.elmTransition = new PointNum(0, 0);

    /**
     * Same as elmTransition but for dragged.
     */
    this.draggedPositionOffset = new PointNum(0, 0);

    this.draggedTransition = new PointNum(0, 0);

    this.isParentLocked = false;
    this.infiniteTransformCount = 0;
  }

  resetTransformCount() {
    if (this.infiniteTransformCount > 0) this.infiniteTransformCount = 0;
  }

  private throwOnInfiniteTransformation(id: string) {
    if (this.infiniteTransformCount > 99) {
      throw new Error(
        `Element ${id} is being transformed endlessly. This is causing infinite recursion affecting the element updater.`
      );
    }
  }

  private setDistanceBtwPositions(
    element: INode,
    axis: Axis,
    elmDirection: Direction
  ) {
    const { occupiedPosition, draggedElm } = this.draggable;

    const positionDiff = Math.abs(
      element.currentPosition[axis] - occupiedPosition[axis]
    );

    const rectDiff = element.getRectDiff(draggedElm, axis);

    if (elmDirection === -1) {
      this.draggedTransition[axis] = positionDiff + rectDiff;
      this.draggedPositionOffset[axis] = rectDiff;

      this.elmTransition[axis] = positionDiff;
    } else {
      this.draggedTransition[axis] = positionDiff;
      this.draggedPositionOffset[axis] = 0;

      this.elmTransition[axis] = positionDiff - rectDiff;
    }
  }

  private updateDraggable(element: INode, elmDirection: Direction) {
    const { currentPosition, grid } = element;

    this.draggable.occupiedPosition.setAxes(
      currentPosition.x + this.draggedPositionOffset.x,
      currentPosition.y + this.draggedPositionOffset.y
    );

    const draggedDirection = -1 * elmDirection;

    this.draggable.occupiedTranslate.increase(
      this.draggedTransition.getMultiplied(draggedDirection)
    );

    this.draggable.gridPlaceholder.clone(grid);
  }

  private updateIndicators(
    element: INode,
    axis: Axis,
    elmDirection: Direction
  ) {
    this.elmTransition.setAxes(0, 0);
    this.draggedTransition.setAxes(0, 0);
    this.draggedPositionOffset.setAxes(0, 0);

    this.setDistanceBtwPositions(element, axis, elmDirection);
    this.updateDraggable(element, elmDirection);
  }

  protected updateDraggedThresholdPosition(x: number, y: number) {
    const {
      threshold,
      draggedElm: { id, offset },
    } = this.draggable;

    threshold.setMainThreshold(id, {
      width: offset.width,
      height: offset.height,
      left: x,
      top: y,
    });
  }

  private addDraggedOffsetToElm(position: IPointAxes, elm: INode, axis: Axis) {
    const rectType = Node.getRectByAxis(axis);

    const { draggedElm } = this.draggable;

    // This initiation needs to append dragged rect based on targeted axis.
    position[axis] += draggedElm.offset[rectType];

    const rectDiff = elm.offset[rectType] - draggedElm.offset[rectType];

    position[axis] += rectDiff;
  }

  /**
   * It calculates the new translate of the dragged element along with grid
   * position inside the container.
   */
  protected getComposedOccupiedTranslateAndGrid(
    SK: string,
    insertAt: number,
    insertFromTop: boolean,
    axis: Axis
  ) {
    const { isEmpty, isOrphan, position, elm, prevElm } =
      store.getInsertionELmMeta(insertAt, SK);

    const { draggedElm, migration, containersTransition } = this.draggable;

    // Getting diff with `currentPosition` includes the element transition
    // as well.
    const composedTranslate = {
      x: Node.getDistance(position, draggedElm, "x"),
      y: Node.getDistance(position, draggedElm, "y"),
    };

    const composedGrid = new PointNum(1, 1);

    // Get the stored position if the branch is empty.
    if (!isEmpty) {
      if (!isOrphan) {
        const { grid } = elm!;

        composedGrid.clone(grid);
      }

      if (insertFromTop) {
        // Don't decrease the first element.
        if (composedGrid[axis] - 1 >= 1) composedGrid[axis] -= 1;
      } else {
        composedGrid[axis] += 1;

        const { marginBottom: mb, marginTop: mt } = migration.prev();

        this.addDraggedOffsetToElm(composedTranslate, elm!, axis);
        composedTranslate[axis] += !isOrphan
          ? Node.getDisplacement(position, prevElm!, axis)
          : typeof mt === "number"
          ? mt
          : typeof mb === "number"
          ? mb
          : containersTransition.margin;
      }
    }

    this.updateDraggedThresholdPosition(
      composedTranslate.x,
      composedTranslate.y
    );

    return { translate: composedTranslate, grid: composedGrid };
  }

  protected getComposedOccupiedPosition(SK: string, axis: Axis) {
    const distLst = store.getElmBranchByKey(SK);

    const { length } = distLst;

    const {
      isEmpty,
      isOrphan,
      position,
      isRestoredLastPosition,
      elm,
      prevElm,
    } = store.getInsertionELmMeta(length - 1, SK);

    // Get the stored position if the branch is empty.
    if (isEmpty) {
      if (!isRestoredLastPosition) {
        throw new Error(
          "Transformation into an empty container in not supported yet."
        );
      }

      return position;
    }

    // The essential insertion position is the last element in the container
    // but also on some cases it's different from retrieved position.
    const composedPosition = elm!.currentPosition.getInstance();

    this.addDraggedOffsetToElm(composedPosition, elm!, axis);

    const { migration, containersTransition } = this.draggable;

    const { marginBottom: mb, marginTop: mt } = migration.latest();

    // Give the priority to the destination first then check the origin.
    const marginBottom = isRestoredLastPosition
      ? Node.getDisplacement(position, elm!, axis)
      : !isOrphan
      ? Node.getDisplacement(position, prevElm!, axis)
      : typeof mt === "number"
      ? mt
      : typeof mb === "number"
      ? mb
      : containersTransition.margin; // orphan to orphan.

    composedPosition[axis] += Math.abs(marginBottom);

    return composedPosition;
  }

  /**
   * Updates element instance and calculates the required transform distance. It
   * invokes for each eligible element in the parent container.
   */
  protected updateElement(id: string, isIncrease: boolean) {
    this.infiniteTransformCount += 1;

    this.throwOnInfiniteTransformation(id);

    const element = store.registry[id];

    const {
      keys: { SK },
    } = element;

    const { grid: siblingsGrid } = store.containers[SK];

    const isContainerHasCol =
      this.draggable.gridPlaceholder.x + 1 <= siblingsGrid.x;

    let axis: Axis;

    if (!isContainerHasCol) {
      axis = "y";

      // const isContainerHasRowAbove =
      //   this.draggable.gridPlaceholder.y + 1 <= siblingsGrid.y;

      // if (isContainerHasRowAbove) {
      //   // Bi-directional
      //   axis = "z";
      // }
    } else {
      axis = "x";
    }

    const elmDirection: Direction = isIncrease ? -1 : 1;

    this.updateIndicators(element, axis, elmDirection);

    // TODO: always true for the first element
    if (!this.isParentLocked) {
      /**
       * By updating the dragged translate, we guarantee that dragged
       * transformation will not triggered until dragged is over threshold
       * which will be detected by isDraggedOutPosition.
       *
       * However, this is only effective when dragged is fit in its new
       * translate.
       *
       * And we have new translate only once. The first element matched the
       * condition is the breaking point element.
       */

      const {
        currentPosition: { x, y },
      } = element;

      this.updateDraggedThresholdPosition(x, y);
    }

    emitInteractiveEvent("onDragOver", element);

    const { migration } = this.draggable;

    /**
     * Start transforming process
     */
    element.setPosition(
      store.getElmBranchByKey(migration.latest().SK),
      elmDirection,
      this.elmTransition,
      migration.latest().id,
      axis
    );

    emitInteractiveEvent("onDragLeave", element);
  }
}

export default DFlexUpdater;
