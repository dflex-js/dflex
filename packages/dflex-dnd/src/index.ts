export { exportedStoreSingleton as store } from "./store";

export { default as DnD } from "./DnD";

export type { DFlexDnDOpts } from "./types";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexLayoutStateEvent,
  DFlexElmMutationEvent,
  DFlexListenerEvents,
} from "./LayoutManager";

export type {
  DFlexScrollContainer,
  DFlexSerializedElement,
  DFlexDOMGenOrder,
} from "@dflex/core-instance";

export type { RegisterInputOpts, DFlexGlobalConfig } from "@dflex/store";

export type {
  AxesPoint,
  AbstractBoxRect as BoxRectAbstract,
  PointNum,
  AbstractBox,
  Dimensions,
} from "@dflex/utils";
