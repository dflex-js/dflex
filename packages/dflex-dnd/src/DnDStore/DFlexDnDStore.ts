import Store from "@dflex/store";
import type { RegisterInputOpts } from "@dflex/store";

import { Tracker, Scroll, canUseDOM } from "@dflex/utils";
import type { Dimensions, ITracker } from "@dflex/utils";

import { DFlexContainer } from "@dflex/core-instance";
import type { IDFlexContainer } from "@dflex/core-instance";

import type { ElmTree, IDFlexDnDStore } from "./types";

import initDFlexListeners from "./DFlexListeners";

import {
  updateBranchVisibility,
  updateElementVisibility,
} from "./DFlexVisibilityUpdater";

import { MAX_NUM_OF_SIBLINGS_BEFORE_DYNAMIC_VISIBILITY } from "./constants";

function throwElementIsNotConnected(id: string) {
  throw new Error(
    `Elements in the branch are not valid. Trying to validate element with id:${id} but failed.\n` +
      `Did you forget to call store.unregister(${id}) or add parenID when register the element?`
  );
}

class DnDStoreImp extends Store implements IDFlexDnDStore {
  containers: Map<string, IDFlexContainer>;

  unifiedContainerDimensions: Map<number, Dimensions>;

  tracker: ITracker;

  observer: MutationObserver | null;

  listeners: ReturnType<typeof initDFlexListeners>;

  private _isDOM: boolean;

  private _isInitialized: boolean;

  constructor() {
    super();
    this.containers = new Map();
    this.unifiedContainerDimensions = new Map();
    this.tracker = new Tracker();
    this._isInitialized = false;
    this._isDOM = false;
    this.observer = null;
    this.listeners = initDFlexListeners();
    this.listeners.notify({ layoutState: "pending", type: "layoutState" });
  }

  private _initWhenRegister() {
    window.onbeforeunload = this.dispose();
  }

  initSiblingContainer(SK: string) {
    if (!this.containers.has(SK)) {
      this.containers.set(SK, new DFlexContainer());
    }

    const branch = this.DOMGen.getElmBranchByKey(SK);

    const firstElemID = branch[0];
    const lastElemID = branch[branch.length - 1];
    const hasSiblings = branch.length > 1;

    if (__DEV__) {
      if (firstElemID && this.registry.get(firstElemID)!.isInitialized) {
        const isHeadNotConnected = !this.registry
          .get(firstElemID)!
          .isConnected();
        let isNotConnected = isHeadNotConnected;

        if (hasSiblings && lastElemID.length > 0) {
          const isTailNotConnected = !this.registry
            .get(lastElemID!)!
            .isConnected();
          isNotConnected = isTailNotConnected || isHeadNotConnected;
        }

        if (isNotConnected) {
          const container = this.containers.get(SK)!;

          if (container.scroll) {
            container.scroll.destroy();
            // @ts-expect-error
            container.scroll = null;
          }

          throwElementIsNotConnected(firstElemID);

          return;
        }
      }
    }

    if (this.containers.get(SK)!.scroll) {
      return;
    }

    const scroll = new Scroll({
      element: this.registry.get(firstElemID)!.ref!,
      requiredBranchKey: SK,
      scrollEventCallback: null,
    });

    // Override allowDynamicVisibility taking into consideration the length of
    // the branch itself. Iterate for a limited number of elements won't be a problem.
    if (
      hasSiblings &&
      scroll.allowDynamicVisibility &&
      this.DOMGen.getElmBranchByKey(SK).length <=
        MAX_NUM_OF_SIBLINGS_BEFORE_DYNAMIC_VISIBILITY
    ) {
      scroll.allowDynamicVisibility = false;
    }

    const container = this.containers.get(SK)!;

    container.scroll = scroll;

    if (scroll.allowDynamicVisibility) {
      scroll.scrollEventCallback = updateBranchVisibility.bind(null, this);
    }
  }

  private _initElmInstance(id: string) {
    const elm = this.registry.get(id)!;

    const {
      keys: { SK },
      depth,
    } = elm;

    const container = this.containers.get(SK)!;

    if (elm.isPaused) {
      elm.resume(container.scroll.scrollX, container.scroll.scrollY);

      if (!elm.isInitialized) {
        return;
      }
    }

    // Using element grid zero to know if the element has been initiated inside
    // container or not.
    if (elm.grid.x === 0) {
      const { offset } = elm;

      container.registerNewElm(
        offset,
        this.unifiedContainerDimensions.get(depth)!
      );

      elm.grid.clone(container.grid);
    }

    updateElementVisibility(elm, container.scroll, false);
  }

  register(element: RegisterInputOpts) {
    if (!this._isDOM) {
      this._isDOM = canUseDOM();

      if (!this._isDOM) return;
    }

    /**
     * If element already exist in the store, then the reattach the reference.
     */

    if (!this._isInitialized) {
      this._initWhenRegister();
      this._isInitialized = true;
    }

    const { id } = element;

    if (this.registry.has(id)) {
      const elm = this.registry.get(id)!;
      if (elm.isInitialized) {
        elm.attach();

        if (elm.isVisible) {
          // Preserves last changes.
          elm.transformElm();
        }
      }

      return;
    }

    const coreInput = {
      id,
      isInitialized: element.priority === "high",
      parentID: element.parentID,
      depth: element.depth || 0,
      readonly: !!element.readonly,
    };

    queueMicrotask(() => {
      const {
        depth,
        keys: { SK },
      } = this.registry.get(id)!;

      if (!this.containers.has(SK)) {
        this.initSiblingContainer(SK);

        if (!this.unifiedContainerDimensions.has(depth)) {
          this.unifiedContainerDimensions.set(depth, {
            width: 0,
            height: 0,
          });
        }
      }

      this._initElmInstance(id!);
    });

    super.register(coreInput);
  }

  getElmSiblingsById(id: string) {
    const element = this.registry.get(id)!;

    if (!element) return null;

    const {
      keys: { SK },
    } = element;

    const siblings = this.getElmBranchByKey(SK);

    return siblings;
  }

  /**
   * Gets element connections instance for a given id.
   *
   * @param id -
   */
  getElmTreeById(id: string): ElmTree {
    const element = this.registry.get(id)!;

    const {
      keys: { SK, PK },
      order,
    } = element;

    /**
     * getting connected branches
     */
    const siblings = this.getElmBranchByKey(SK);
    const parents = this.getElmBranchByKey(PK);

    /**
     * getting parent instance
     */
    let parent = null;
    if (parents !== undefined) {
      const parentsID = parents[order.parent];
      parent = this.registry.get(parentsID)!;
    }

    return {
      element,
      parent,

      branches: {
        siblings,
        parents,
      },
    };
  }

  getSerializedElm(id: string) {
    if (__DEV__) {
      if (!this.registry.has(id)) {
        throw new Error(
          `getSerializedElm: Element with id ${id} does not exist in the registry.`
        );
      }
    }

    return this.registry.has(id) ? this.registry.get(id)!.exportJSON() : null;
  }

  private _clearBranchesScroll() {
    this.DOMGen.forEachBranch((SK) => {
      const container = this.containers.get(SK)!;
      if (container.scroll) {
        container.scroll.destroy();
      }
    });
  }

  /**
   * Unregister DnD element.
   *
   * Note: This will remove the element registry and the branch array. But,
   * in case all the branches will be removed.
   * This means, if, in rare cases when the user removes one element and keeps
   * the rest this methods going to generate a bug. It's going to remove an
   * element without updating the indexes inside registry instances.
   *
   * @param id -
   *
   */
  unregister(id: string) {
    const {
      keys: { SK },
      order: { self },
    } = this.registry.get(id)!;

    this.DOMGen.removeElmIDFromBranch(SK, self);

    super.unregister(id);

    // Nothing left?
    // Reset the branch instances.
    if (this.DOMGen.getElmBranchByKey(SK).length === 0) {
      this._clearBranchesScroll();
    }
  }

  dispose() {
    if (!this._isInitialized) return null;

    this._isInitialized = false;

    return null;
  }

  destroy() {
    this.dispose();

    this._clearBranchesScroll();

    // Destroys all registered instances.
    super.destroy();

    if (this.observer) {
      this.observer.disconnect();
    }
  }
}
declare global {
  // eslint-disable-next-line
  var $DFlex: DnDStoreImp;
}
export default (function createStoreInstance() {
  const store = new DnDStoreImp();

  if (__DEV__) {
    if (canUseDOM()) {
      if (!globalThis.$DFlex) {
        globalThis.$DFlex = store;
      } else {
        throw new Error("DFlex store instances is already defined.");
      }
    }
  }

  return store;
})();
