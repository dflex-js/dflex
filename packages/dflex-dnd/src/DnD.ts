import type { AxesPoint } from "@dflex/utils";

import DraggableInteractive from "./Draggable";
import Droppable from "./Droppable";
import { scheduler, store } from "./LayoutManager";

import type { DFlexDnDOpts, FinalDndOpts } from "./types";

import { extractOpts } from "./utils/extractOpts";
import { defaultOpts } from "./utils/constants";

class DnD extends Droppable {
  /**
   *
   * @param id -
   * @param initCoordinates -
   * @param opts -
   */
  constructor(
    id: string,
    initCoordinates: AxesPoint,
    opts: DFlexDnDOpts = defaultOpts
  ) {
    if (__DEV__) {
      if (!store.registry.has(id)) {
        throw new Error(`DFlex: ${id} is not registered in the Store.`);
      }
    }

    const options = extractOpts(opts);

    const draggable = new DraggableInteractive(
      id,
      initCoordinates,
      options as FinalDndOpts
    );

    super(draggable);

    scheduler(store, null, null, { layoutState: "ready", type: "layoutState" });
  }
}

export default DnD;
