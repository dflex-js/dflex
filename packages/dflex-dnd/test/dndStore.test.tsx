/* eslint-disable import/no-extraneous-dependencies */
import * as React from "react";
import * as ReactTestUtils from "react-dom/test-utils";
import { Root, createRoot } from "react-dom/client";

import { store } from "../src/DnDStore";
import { getInsertionELmMeta } from "../src/Droppable/DFlexUpdater";

describe("DnD Store", () => {
  const elm1 = {
    id: "id-1",
    depth: 0,
    parentID: "",
    readonly: false,
  };

  const elm2 = {
    id: "id-2",
    depth: 0,
    parentID: "",
    readonly: false,
  };

  const elm3 = {
    id: "id-3",
    depth: 0,
    parentID: "",
    readonly: false,
  };

  const elm4 = {
    id: "id-4",
    depth: 0,
    parentID: "",
    readonly: false,
  };

  let container: HTMLDivElement | null;
  let reactRoot: Root;

  const positions: { [k: string]: DOMRect } = {};

  beforeAll(() => {
    container = document.createElement("div");
    reactRoot = createRoot(container);
    document.body.appendChild(container);

    function render(elm: typeof elm3) {
      const ref = React.createRef<HTMLDivElement>();

      React.useEffect(() => {
        if (ref.current) {
          store.register(elm);

          positions[elm.id] = ref.current.getBoundingClientRect();
        }

        return () => {
          store.unregister(elm.id);
        };
      }, [ref]);

      return <div id={elm.id} key={elm.id} ref={ref} />;
    }

    function init() {
      function TestBase() {
        return <div>{[elm1, elm2, elm3, elm4].map((elm) => render(elm))}</div>;
      }

      ReactTestUtils.act(() => {
        reactRoot.render(<TestBase />);
      });
    }

    init();
  });

  afterAll(() => {
    document.body.removeChild(container!);
    container = null;
  });

  describe("Shallow test the store", () => {
    it("Element is initiated", () => {
      store.register(elm1);
      expect(store.registry).toMatchSnapshot();
    });

    it("Returns element branch and parent", () => {
      const { branches, parent } = store.getElmTreeById(elm1.id);

      expect(branches).toMatchInlineSnapshot(`
        Object {
          "parents": Array [],
          "siblings": Array [
            "id-1",
            "id-2",
            "id-3",
            "id-4",
          ],
        }
      `);

      expect(parent).toMatchInlineSnapshot(`undefined`);
    });
  });

  describe("getInsertionELmMeta", () => {
    const preservePosition = {
      x: 1000,
      y: 1000,
    };

    const originLength = 4;

    let SK: string;

    describe("Normal branch without injection", () => {
      it("Checks orphan branch", () => {
        ({
          element: {
            keys: { SK },
          },
        } = store.getElmTreeById(elm1.id));

        const { elm, ...rest } = getInsertionELmMeta(0, SK);

        expect(rest).toMatchInlineSnapshot(`
          Object {
            "isEmpty": false,
            "isOrphan": false,
            "isRestoredLastPosition": false,
            "position": PointNum {
              "x": 0,
              "y": 0,
            },
            "prevElm": null,
          }
        `);
      });

      it("Register three elements in the store", () => {
        store.register(elm2);
        store.register(elm3);
        store.register(elm4);

        // @ts-ignore - this is a test
        expect(store.DOMGen.getElmBranchByKey(SK)).toStrictEqual([
          "id-1",
          "id-2",
          "id-3",
          "id-4",
        ]);
      });

      it("Checks the last element", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(3, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": false,
                  }
              `);

        const pos = positions[elm4.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm4.id);
        expect(prevElm!.id).toBe(elm3.id);
      });

      it("Preserve the last element", () => {
        const elmContainer = store.containers.get(SK)!;

        elmContainer.preservePosition(preservePosition);
        elmContainer.originLength = originLength;
      });

      it("Restores the preserved position when calling for last element", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(3, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        expect(position.getInstance()).toStrictEqual(preservePosition);

        expect(elm!.id).toBe(elm4.id);
        expect(prevElm!.id).toBe(elm3.id);
      });

      it("Restores the position even preserved is defined because it's not the last element", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(2, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = positions[elm3.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm3.id);
        expect(prevElm!.id).toBe(elm2.id);
      });

      it("Restores the position for the first element with correct flags", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(0, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = positions[elm1.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm1.id);
        expect(prevElm).toBeNull();
      });
    });

    describe("Injected branch with empty string", () => {
      it("Inject empty string", () => {
        // @ts-ignore - this is a test
        store.DOMGen.addElmIDToBranch(SK, "");

        // @ts-ignore - this is a test
        expect(store.DOMGen.getElmBranchByKey(SK)).toStrictEqual([
          "id-1",
          "id-2",
          "id-3",
          "id-4",
          "",
        ]);
      });

      it("Restores the the last element when the branch is extended", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(4, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": false,
                  }
              `);

        const pos = positions[elm4.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm4.id);
        expect(prevElm!.id).toBe(elm3.id);
      });

      it("Restores the position even preserved is defined because it's not the last element", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(3, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = positions[elm4.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm4.id);
        expect(prevElm!.id).toBe(elm3.id);
      });

      it("Restores the position for the first element with correct flags", () => {
        const { elm, prevElm, position, ...rest } = getInsertionELmMeta(0, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = positions[elm1.id];

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm!.id).toBe(elm1.id);
        expect(prevElm).toBeNull();
      });
    });
  });
});
