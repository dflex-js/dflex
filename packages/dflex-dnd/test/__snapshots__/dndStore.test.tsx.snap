// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DnD Store Shallow test the store Element is initiated 1`] = `
Object {
  "id-1": DFlexNode {
    "_hasAttribute": Object {},
    "animatedFrame": 1,
    "currentPosition": PointNum {
      "x": 0,
      "y": 0,
    },
    "depth": 0,
    "grid": PointNum {
      "x": 1,
      "y": 1,
    },
    "hasToTransform": false,
    "id": "id-1",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "offset": Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 0,
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "readonly": false,
    "ref": <div
      id="id-1"
    />,
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
  "id-2": DFlexNode {
    "_hasAttribute": Object {},
    "animatedFrame": null,
    "currentPosition": PointNum {
      "x": 0,
      "y": 0,
    },
    "depth": 0,
    "grid": PointNum {
      "x": 1,
      "y": 1,
    },
    "hasToTransform": false,
    "id": "id-2",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "offset": Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 0,
    },
    "order": Object {
      "parent": 0,
      "self": 1,
    },
    "readonly": false,
    "ref": <div
      id="id-2"
    />,
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
  "id-3": DFlexNode {
    "_hasAttribute": Object {},
    "animatedFrame": null,
    "currentPosition": PointNum {
      "x": 0,
      "y": 0,
    },
    "depth": 0,
    "grid": PointNum {
      "x": 1,
      "y": 1,
    },
    "hasToTransform": false,
    "id": "id-3",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "offset": Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 0,
    },
    "order": Object {
      "parent": 0,
      "self": 2,
    },
    "readonly": false,
    "ref": <div
      id="id-3"
    />,
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
  "id-4": DFlexNode {
    "_hasAttribute": Object {},
    "animatedFrame": null,
    "currentPosition": PointNum {
      "x": 0,
      "y": 0,
    },
    "depth": 0,
    "grid": PointNum {
      "x": 1,
      "y": 1,
    },
    "hasToTransform": false,
    "id": "id-4",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "offset": Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 0,
    },
    "order": Object {
      "parent": 0,
      "self": 3,
    },
    "readonly": false,
    "ref": <div
      id="id-4"
    />,
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
}
`;
